service/history/replication/executable_sync_hsm_task.go

```diff
@@ -0,0 +1,241 @@
+		e.WorkflowID,
+	)
+	if err != nil {
+		return err
+	}
+	engine, err := shardContext.GetEngine(ctx)
+	if err != nil {
+		return err
+	}
+	return engine.SyncHSM(ctx, &shard.SyncHSMRequest{
+		WorkflowKey:         e.WorkflowKey,
+		StateMachineNode:    e.taskAttr.StateMachineNode,
+		EventVersionHistory: e.taskAttr.VersionHistory,
+	})
+
+}
+
+func (e *ExecutableSyncHSMTask) HandleErr(err error) error {
+	switch retryErr := err.(type) {
+	case nil, *serviceerror.NotFound:
+		return nil
+	case *serviceerrors.RetryReplication:
+		namespaceName, _, nsError := e.GetNamespaceInfo(headers.SetCallerInfo(
+			context.Background(),
+			headers.SystemPreemptableCallerInfo,
+		), e.NamespaceID)
+		if nsError != nil {
+			return err
+		}
+		ctx, cancel := newTaskContext(namespaceName, e.Config.ReplicationTaskApplyTimeout())
+		defer cancel()
+
+		if doContinue, resendErr := e.Resend(
+			ctx,
+			e.ExecutableTask.SourceClusterName(),
+			retryErr,
+			ResendAttempt,
+		); resendErr != nil || !doContinue {
+			return err
+		}
+		return e.Execute()
+	default:
+		e.Logger.Error("Sync HSM replication task encountered error",
+			tag.WorkflowNamespaceID(e.NamespaceID),
+			tag.WorkflowID(e.WorkflowID),
+			tag.WorkflowRunID(e.RunID),
+			tag.TaskID(e.ExecutableTask.TaskID()),
+			tag.Error(err),
+		)
+		return err
+	}
+}
+
+func (e *ExecutableSyncHSMTask) MarkPoisonPill() error {
+
+	replicationTaskInfo := e.toReplicationTaskInfo()
+
+	if e.markPoisonPillAttempts >= MarkPoisonPillMaxAttempts {
+		e.Logger.Error("MarkPoisonPill reached max attempts",
+			tag.SourceCluster(e.SourceClusterName()),
+			tag.ReplicationTask(replicationTaskInfo),
+			tag.NewAnyTag("sync-hsm-task-attr", e.taskAttr),
+		)
+		return nil
+	}
+	e.markPoisonPillAttempts++
+
+	shardContext, err := e.ShardController.GetShardByNamespaceWorkflow(
+		namespace.ID(e.NamespaceID),
+		e.WorkflowID,
+	)
+	if err != nil {
+		return err
+	}
+
+	e.Logger.Error("enqueue sync HSM replication task to DLQ",
+		tag.ShardID(shardContext.GetShardID()),
+		tag.WorkflowNamespaceID(e.NamespaceID),
+		tag.WorkflowID(e.WorkflowID),
+		tag.WorkflowRunID(e.RunID),
+		tag.TaskID(e.ExecutableTask.TaskID()),
+	)
+
+	ctx, cancel := newTaskContext(e.NamespaceID, e.Config.ReplicationTaskApplyTimeout())
+	defer cancel()
+
+	// TODO: GetShardID will break GetDLQReplicationMessages we need to handle DLQ for cross shard replication.
+	return writeTaskToDLQ(ctx, e.DLQWriter, shardContext, e.SourceClusterName(), replicationTaskInfo)
+}
+
+func (e *ExecutableSyncHSMTask) toReplicationTaskInfo() *persistencespb.ReplicationTaskInfo {
+	return &persistencespb.ReplicationTaskInfo{
+		NamespaceId:    e.NamespaceID,
+		WorkflowId:     e.WorkflowID,
+		RunId:          e.RunID,
+		TaskType:       enumsspb.TASK_TYPE_REPLICATION_SYNC_HSM,
+		TaskId:         e.ExecutableTask.TaskID(),
+		VisibilityTime: timestamppb.New(e.TaskCreationTime()),
+	}
+}
+
+// TODO: implement the following methods to batch syncHSM task if needed
```

Comment:
IIUC, this should be batchable by merging two tasks and overwriting the version history + HSM tree.

Not critical since this is a temporary solution anyways.

Reply:
Yeah, we can blindly take the tree from the task with higher taskID since replication tasks are converted in fifo order in source cluster. 
We can add batching logic later if needed, but like you said it's not critical.